flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableTest.java

```diff
@@ -238,8 +208,112 @@ public void testGenerateSlotReport() throws SlotNotFoundException {
+			assertThat(allocatedSlots.hasNext(), is(false));
+		}
+	}
+
+	@Test
+	public void testAddTask() throws Exception {
+		final JobID jobId = new JobID();
+		final ExecutionAttemptID executionAttemptId = new ExecutionAttemptID();
+		final AllocationID allocationId = new AllocationID();
+		Task task = TaskSlotUtils.createTask(
+			jobId,
+			executionAttemptId,
+			allocationId,
+			EmptyInvokable.class);
+		try (final TaskSlotTable taskSlotTable =
+				 createTaskSlotTableWithStartedTask(task, new TestingSlotActionsBuilder().build())) {
+			Iterator<Task> tasks = taskSlotTable.getTasks(jobId);
+			Task nextTask = tasks.next();
+			assertThat(nextTask.getExecutionId(), is(executionAttemptId));
+			assertThat(nextTask.getAllocationId(), is(allocationId));
+			assertThat(tasks.hasNext(), is(false));
+		}
+	}
+
+	@Test(timeout = 10000)
+	public void testRemoveTaskCallsFreeSlotAction() throws Exception {
+		final JobID jobId = new JobID();
+		final ExecutionAttemptID executionAttemptId = new ExecutionAttemptID();
+		final AllocationID allocationId = new AllocationID();
+		CompletableFuture<AllocationID> freeSlotFuture = new CompletableFuture<>();
+		SlotActions slotActions = new TestingSlotActions(freeSlotFuture::complete, (aid, uid) -> {});
+		Task task = TaskSlotUtils.createTask(
+			jobId,
+			executionAttemptId,
+			allocationId,
+			EmptyInvokable.class);
+		try (final TaskSlotTable taskSlotTable = createTaskSlotTableWithStartedTask(task, slotActions)) {
+			// we have to initiate closing of the slot externally
+			// to enable that the last remaining finished task does the final slot freeing
+			taskSlotTable.freeSlot(allocationId);
+			taskSlotTable.removeTask(executionAttemptId);
+			assertThat(freeSlotFuture.get(), is(allocationId));
+		}
+	}
+
+	@Test(timeout = 10000)
+	public void testFreeSlotInterruptsSubmittedTask() throws Exception {
+		final JobID jobId = new JobID();
+		final ExecutionAttemptID executionAttemptId = new ExecutionAttemptID();
+		final AllocationID allocationId = new AllocationID();
+		TestInterruptableInvokable.reset();
+		Task task = TaskSlotUtils.createTask(
+			jobId,
+			executionAttemptId,
+			allocationId,
+			TestInterruptableInvokable.class);
+		final TaskSlotTable taskSlotTable = createTaskSlotTableWithStartedTask(task, new TestingSlotActionsBuilder().build());
+		TestInterruptableInvokable.waitForStarted();
+		assertThat(taskSlotTable.freeSlot(allocationId), is(-1));
+		TestInterruptableInvokable.waitForInterrupted();
+		CompletableFuture<Void> closingFuture = taskSlotTable.closeAsync();
+		assertThat(closingFuture.isDone(), is(false));
+		TestInterruptableInvokable.finish();
+		closingFuture.get();
+	}
+
+	private static TaskSlotTableImpl createTaskSlotTableAndStart(int numberOfSlots) {
+		final TaskSlotTableImpl taskSlotTable = TaskSlotUtils.createTaskSlotTable(numberOfSlots);
+		taskSlotTable.start(new TestingSlotActionsBuilder().build(), Executors.directExecutor());
```

Comment:
Using the `directExecutor` can be problematic as it allows for thread interleavings which are not possible with a proper `Executor`. I assume that in this case it is not problematic, though, because we use the executor only for the clean up call.